src/icon_dearchiver.cc

```diff
@@ -0,0 +1,480 @@
+  { RESOURCE_TYPE_ih32, 9216, true,  13 },
+  { RESOURCE_TYPE_h8mk, 2304, false, 14 },
+};
+static_assert(sizeof(ICON_TYPES) / sizeof(ICON_TYPES[0]) == ICON_TYPE_COUNT);
+
+static constexpr uint8_t icns_type_to_icns_idx(uint32_t icns_type) {
+  for (uint8_t i = 0; i < ICON_TYPE_COUNT; ++i) {
+    if (ICON_TYPES[i].icns_type == icns_type) {
+      return i;
+    }
+  }
+  throw logic_error("Unsupported icns icon type");
+}
+
+// Order must match the one in `ICON_TYPES` above, use helper function to
+// guarantee this
+static constexpr uint8_t ICON_TYPE_ICNN = icns_type_to_icns_idx(RESOURCE_TYPE_ICNN);
+static constexpr uint8_t ICON_TYPE_icl4 = icns_type_to_icns_idx(RESOURCE_TYPE_icl4);
+static constexpr uint8_t ICON_TYPE_icl8 = icns_type_to_icns_idx(RESOURCE_TYPE_icl8);
+static constexpr uint8_t ICON_TYPE_il32 = icns_type_to_icns_idx(RESOURCE_TYPE_il32);
+static constexpr uint8_t ICON_TYPE_l8mk = icns_type_to_icns_idx(RESOURCE_TYPE_l8mk);
+
+static constexpr uint8_t ICON_TYPE_icsN = icns_type_to_icns_idx(RESOURCE_TYPE_icsN);
+static constexpr uint8_t ICON_TYPE_ics4 = icns_type_to_icns_idx(RESOURCE_TYPE_ics4);
+static constexpr uint8_t ICON_TYPE_ics8 = icns_type_to_icns_idx(RESOURCE_TYPE_ics8);
+static constexpr uint8_t ICON_TYPE_is32 = icns_type_to_icns_idx(RESOURCE_TYPE_is32);
+static constexpr uint8_t ICON_TYPE_s8mk = icns_type_to_icns_idx(RESOURCE_TYPE_s8mk);
+
+static constexpr uint8_t ICON_TYPE_ichN = icns_type_to_icns_idx(RESOURCE_TYPE_ichN);
+static constexpr uint8_t ICON_TYPE_ich4 = icns_type_to_icns_idx(RESOURCE_TYPE_ich4);
+static constexpr uint8_t ICON_TYPE_ich8 = icns_type_to_icns_idx(RESOURCE_TYPE_ich8);
+static constexpr uint8_t ICON_TYPE_ih32 = icns_type_to_icns_idx(RESOURCE_TYPE_ih32);
+static constexpr uint8_t ICON_TYPE_h8mk = icns_type_to_icns_idx(RESOURCE_TYPE_h8mk);
+
+// .icns files must contain the icons in a specific order, namely b/w icons
+// last, or they don't show up correctly in Finder
+// TODO: system-made .icns don't do this?
+static constexpr uint8_t ICON_ICNS_ORDER[] = {
+ ICON_TYPE_ics4,
+ ICON_TYPE_ics8,
+ ICON_TYPE_is32,
+ ICON_TYPE_s8mk,
+ ICON_TYPE_icl4,
+ ICON_TYPE_icl8,
+ ICON_TYPE_il32,
+ ICON_TYPE_l8mk,
+ ICON_TYPE_ich4,
+ ICON_TYPE_ich8,
+ ICON_TYPE_ih32,
+ ICON_TYPE_h8mk,
+ ICON_TYPE_icsN,
+ ICON_TYPE_ICNN,
+ ICON_TYPE_ichN,
+};
+static_assert(sizeof(ICON_ICNS_ORDER) == ICON_TYPE_COUNT * sizeof(ICON_ICNS_ORDER[0]));
+
+
+
+static void unpack_bits(StringReader& in, void* uncompressed_data, uint32_t uncompressed_size) {
+  uint8_t*        out = static_cast<uint8_t*>(uncompressed_data);
+  uint8_t* const  out_end = out + uncompressed_size;
+  while (out < out_end) {
+    int8_t len = in.get_s8();
+    if (len < 0) {
+      // -len+1 repetitions of the next byte
+      uint8_t byte = in.get_u8();
+      for (int i = 0; i < -len + 1; ++i) {
+        *out++ = byte;
+      }
+    } else {
+      // len + 1 raw bytes
+      in.readx(out, len + 1);
+      out += len + 1;
+    }
+  }
+}
```

Comment:
There is another implementation of this which I recently added; I'm going to move it to src/DataCodecs so it's easily reusable. Curiously, the documentation I read when writing that function claims that 0x80 should be treated as a no-op when decoding PackBits, but I don't understand why that would be part of the design (and indeed, your version of the function here doesn't do that). Do you know what the correct behavior for command 0x80 should be?

Reply:
I based my implementation on Technote 171, "Things You Wanted to Know About _PackBits", and blissfully skipped over this paragraph:

> Warning: _PackBits never generates the value -128 ($80) as a flag-counter byte, but a few PackBits-like routines that are built into some applications do. _UnpackBits handles this situation by skipping any flag-counter byte with this value and interpreting the next byte as the next flag-counter byte. If you’re writing your own UnpackBits-like routine, make sure it handles this situation in the same way.

So your implementation is correct.

Curiously, [QuickDraw 1.0](https://computerhistory.org/blog/macpaint-and-quickdraw-source-code/)'s implementation of `UnpackBits` in `Bitmap.a` says:

```
        BVS.S   @3                              ;IGNORE $80 FOR BACKWARD COMPAT
```